Segment appender improvements

mmaslankaprv opened the pull request:
This PR contains a very superficial refactoring of segment appender. The PR  removes old not used code, introduces some changes in appender API, removes typos and adds tests. This work is going to be a foundation for appender fixes and rework of the way we handle buffers. 


<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
s/segment_appender: use format_to method instead of stream operator

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
s/appender: do not pass number of chunks to appender

Segment appender does not need information about number of chunks. The
`chunk_no` value was passed to very old appender implementation.
Currently it wasn't used in any other place than for logging purposes

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
s/appender: simplified callback management in appender

Previously segment appender used a dedicated interface to pass in the
callbacks. Simplified it with a use of noncopyable function as there is
currently only one type of a callback that can be set on the appender.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
s/appender: introduced segment appender stats

Introduced a stats class that contains basic stats about the appender
including total number of operations executed and actual number of disk
writes together with number of fsyncs.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
s/tests: added segment appender test

Added low level test validating segment appender behavior. The test
validates the correctness of an appender by executing operations over
the appender and a simple in memory buffer at the same time. The result
of all operations must be the same i.e. file content that is managed by
the appender must be exactly the same as the in memory iobuf content.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
s/segment_appender: fixed typos in code and comments

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
test_utils/tmpfile: do not log with info level

Signed-off-by: Michał Maślanka <michal@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75101-->
<!--ae4f3a1f-256c-434f-8316-93dd7493fd6e-->
<!--title-->
## Retry command for [Build#75101](https://buildkite.com/redpanda/redpanda/builds/75101)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/tls_version_test.py::TLSVersionTestRSA.test_change_version@{"version":1}
tests/rptest/tests/tls_metrics_test.py::TLSMetricsTestChain.test_cert_chain_metrics
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--ec671cd8-e05d-4bfd-9569-708db8802494-->
<!--TEST RESULTS-->
<!--54632d53-3687-4986-a2c6-70c614431dfc-->
<!--TEST RESULTS-->
<!--d8acc8b5-5594-48ba-a5b8-9e3dfb9cf4bf-->
<!--TEST RESULTS-->
<!--a30e052b-ca71-4640-8c67-eb4610d7d1b8-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75101">build#75101</a></summary>


| test_class                                        | test_method                                       | test_arguments                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                               |
|:--------------------------------------------------|:--------------------------------------------------|:---------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTxRemovalUpgradeTest                 | test_tx_control_batch_removal_with_upgrade        | {"test_case_name": "All aborts"} | integration | https://buildkite.com/redpanda/redpanda/builds/75101#019a2aa3-8918-49a6-9e85-02ca30895ba9 | FLAKY         | 17/21    | upstream reliability is '86.66666666666667'. current run reliability is '80.95238095238095'. drift is 5.71429 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| TLSMetricsTestChain                               | test_cert_chain_metrics                           | null                             | integration | https://buildkite.com/redpanda/redpanda/builds/75101#019a2aa3-8911-4970-b989-3e6784c24155 | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TLSMetricsTestChain&test_method=test_cert_chain_metrics                                  |
| TLSVersionTestRSA                                 | test_change_version                               | {"version": 1}                   | integration | https://buildkite.com/redpanda/redpanda/builds/75101#019a2aa3-8911-4970-b989-3e6784c24155 | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TLSVersionTestRSA&test_method=test_change_version                                        |
| src/v/storage/tests/segment_appender_rpbench_test | src/v/storage/tests/segment_appender_rpbench_test |                                  | unit        | https://buildkite.com/redpanda/redpanda/builds/75101#019a2a71-3a2e-458b-91b5-2d1aad4f1739 | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                                            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75107">build#75107</a></summary>


| test_class                                        | test_method                                       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                           |
|:--------------------------------------------------|:--------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkUpdateBrokersTests                      | test_update_brokers                               | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75107#019a2b39-2631-4c9b-9d0c-a8374d3ec684 | FLAKY         | 20/21    | upstream reliability is '96.96969696969697'. current run reliability is '95.23809523809523'. drift is 1.7316 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkUpdateBrokersTests&test_method=test_update_brokers         |
| LogCompactionTxRemovalTest                        | test_tx_control_batch_removal                     | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75107#019a2b39-263c-4b5f-8f6c-b6a25297c241 | FLAKY         | 18/21    | upstream reliability is '94.36619718309859'. current run reliability is '85.71428571428571'. drift is 8.65191 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal |
| TLSMetricsTest                                    | test_services                                     | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75107#019a2b39-2634-47bb-878a-71b61fdde9e2 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TLSMetricsTest&test_method=test_services                             |
| src/v/storage/tests/segment_appender_rpbench_test | src/v/storage/tests/segment_appender_rpbench_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/75107#019a2b06-d0e1-4300-ab90-fedd67d8dc25 | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/storage/tests/segment_appender_test         | src/v/storage/tests/segment_appender_test         |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/75107#019a2b06-d0e1-4300-ab90-fedd67d8dc25 | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75199">build#75199</a></summary>


| test_class                        | test_method                                | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                               |
|:----------------------------------|:-------------------------------------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTxRemovalUpgradeTest | test_tx_control_batch_removal_with_upgrade | {"test_case_name": "All commits"} | integration | https://buildkite.com/redpanda/redpanda/builds/75199#019a3017-edbb-4b69-acb6-ef4be997e92d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| PartitionMovementTest             | test_empty                                 | {"num_to_upgrade": 0}             | integration | https://buildkite.com/redpanda/redpanda/builds/75199#019a3018-4a1d-4dd3-bdf8-190c54e77816 | FLAKY         | 17/21    | upstream reliability is '88.78600823045268'. current run reliability is '80.95238095238095'. drift is 7.83363 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty                                             |
| TxUpgradeCompactionTest           | upgrade_with_compaction_test               | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/75199#019a3017-edb8-4b59-9035-a53d476b11bd | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxUpgradeCompactionTest&test_method=upgrade_with_compaction_test                         |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75213">build#75213</a></summary>


| test_class              | test_method                          | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:------------------------|:-------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest   | test_creating_and_listing_migrations | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/75213#019a3094-9fc7-44db-858c-21842b2d8428 | FLAKY         | 15/21    | upstream reliability is '99.72789115646259'. current run reliability is '71.42857142857143'. drift is 28.29932 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| MountUnmountIcebergTest | test_simple_remount                  | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/75213#019a3094-9fc3-470d-be01-e2ec2281b313 | FLAKY         | 16/21    | upstream reliability is '78.62595419847328'. current run reliability is '76.19047619047619'. drift is 2.43548 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                |
</details>
<!--/body-->

dotnwat commented on src/v/storage/segment_appender.h:
```diff
@@ -275,7 +278,7 @@ class segment_appender {
     // dispatched
     size_t _merged_writes{0};
-    callbacks* _callbacks = nullptr;
+    std::unique_ptr<committed_offset_clb> _committed_offset_clb;
```
is it a unique ptr for its nullability, or because of something else? if the former, i'm pretty sure `noncopyable_function` already provides an `operator bool()` to use to check for it being set or empty?

dotnwat commented on src/v/storage/segment_appender.cc:
```diff
@@ -192,7 +192,9 @@ ss::future<> segment_appender::do_append(const char* buf, size_t n) {
 void segment_appender::check_no_dispatched_writes() {
     vassert(
-      _inflight_dispatched == 0, "Unexpected pending head write {}", *this);
+      _stats.inflight_dispatched == 0,
```
hmm, i wonder if we want to keep some counters outside of the stats structure since they are more than just stats, they are used in the flow control.

dotnwat reviewed (commented):
lgtm pending travis' feedback and a couple questions

mmaslankaprv commented:
> @mmaslankaprv was the bench deleted on purpse?

I deleted it with the intention to submit it in a follow up PR, it was timing out in CI, i need to investigate why

mmaslankaprv committed:
s/appender: fixed race condition when waiting for flush operation

When executing `flush()` segment appender verifies if there are pending
bytes in head and instead of calling flush on the underlying `file` it
dispatches backgroud head write first and waits for flush completion
after that is a result of that background write and advancement of
committed offset.

Previously the appender called the `dispatch_background_head_write()`
method before accessing the future. In some rare situations it might
happened (especially when all the futures that are part of the
background flush are ready) that the operation was already completed and
deleted before its future was returned to the caller.

Now the future is requested first before returning to the caller this
way there is a guarantee that we are not accessing dangling reference.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
s/tests: added very simple appender benchmark

Added very simple tests for benchmarking the segment appender. The test
is not intended to measure the absolute performance of the appender but
rather than that to serve as a reference point when making changes in
segment appender.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
I added then benchmark back together with a fix of an issue leading to a timeout in benchmark test

dotnwat commented on src/v/storage/segment_appender.cc:
```diff
@@ -611,13 +620,17 @@ void segment_appender::dispatch_background_head_write() {
     if (_head && _head->bytes_pending()) {
         auto& w = _flush_ops.emplace_back(file_byte_offset());
+        // get future first, as the flush_op may be deleted/moved when
+        // dispatching background head write.
+        auto f = w.p.get_future();
         dispatch_background_head_write();
-        return w.p.get_future();
+        return f;
```
the code path looks like this is indeed possible, but is there something that is preventing it from ever occurring / being problematic?

mmaslankaprv replied:
In normal code path this never happened as there was always a scheduling point in `dispatch_background_head_write`. The scheduling point is not present when using `tmpbuf` file implementation.

dotnwat reviewed (approved):
🥳
